Code/GraphMol/Depictor/EmbeddedFrag.cpp

```diff
@@ -341,11 +308,118 @@ int EmbeddedFrag::findNeighbor(
   return -1;
 }
 
+void EmbeddedFrag::setupAttachmentPoints() {
+  // now for points that new atoms will be added to later on we need to do some
+  // setup
+  for (auto dai : d_attachPts) {
+    // find the neighbors that are already embedded for each of these atoms
+    RDKit::INT_VECT doneNbrs;
+    const auto &enbrs = d_eatoms[dai].neighs;
+    for (const auto nbrAtom :
+         dp_mol->atomNeighbors(dp_mol->getAtomWithIdx(dai))) {
+      if (std::find(enbrs.begin(), enbrs.end(),
+                    static_cast<int>(nbrAtom->getIdx())) == enbrs.end()) {
+        // we found a neighbor that is part of this embedded system
+        doneNbrs.push_back(nbrAtom->getIdx());
+      }
+    }
+    if (doneNbrs.empty()) {
+      d_eatoms[dai].normal = RDGeom::Point2D(1., 0.);
+      d_eatoms[dai].angle = -1.;
+    } else if (doneNbrs.size() == 1) {
+      auto nbid = doneNbrs.front();
+      d_eatoms[dai].nbr1 = nbid;
+      d_eatoms[dai].normal =
+          computeNormal(d_eatoms[dai].loc, d_eatoms[nbid].loc);
+    } else if (doneNbrs.size() == 2) {
+      auto nb1 = doneNbrs[0];
+      auto nb2 = doneNbrs[1];
+      d_eatoms[dai].nbr1 = nb1;
+      d_eatoms[dai].nbr2 = nb2;
+      d_eatoms[dai].angle =
+          computeAngle(d_eatoms[dai].loc, d_eatoms[nb1].loc, d_eatoms[nb2].loc);
+    } else if (doneNbrs.size() >= 3) {
+      // this is a pain - delegate it to a utility function
+      this->computeNbrsAndAng(dai, doneNbrs);
+    }
+  }
+}
+
+bool EmbeddedFrag::matchToTemplate(const RDKit::INT_VECT &ringSystemAtoms, unsigned int ring_count) {
```

Comment:
General comment that came to mind as I was reading through this function: we should be sure that we have some test molecules which match multiple templates: both in different parts of the molecule (i.e. separated by a linker) and having overlapping matches (assuming that's possible).
We also need tests looking at the interaction between templates and parts of the molecule which have fixed coordinates (like a fixed core)

Reply:
That makes sense -- I added some python tests for this. The overlapping matches should not be possible because the matches are ring system by ring system and must be exact.